### PR TITLE
feat(ui): create stat widget

### PR DESCRIPTION
**Pull request by BillyFigueroa**
## Description
Creating StatWidget which will be responsible for many of the stat components needed across the k8s and other views like the SBOM, CVE etc

## Related Issue

- #384 
- #281 

#### Before
Screenshot 2024-09-24 at 11 44 00 AM

#### After
Screenshot 2024-09-24 at 11 43 26 AM

#### Critical Theme
Screenshot 2024-09-27 at 12 22 00 PM

#### Warning Theme
Screenshot 2024-09-27 at 12 22 29 PM

## Figma
https://www.figma.com/design/zmKcJ9Xin7ChzyGy6RCFLe/UDS-Runtime-(UI%2FCLI)?node-id=4325-5172&node-type=frame&m=dev

### CODE REVIEW

**Comment by UncleGedd** [comment]
Hey @BillyFigueroa this looks great! Could you point me to the design for this?

**Comment by BillyFigueroa** [comment]
> Hey @BillyFigueroa this looks great! Could you point me to the design for this?

@UncleGedd thanks. Here are 2 examples. I will likely add some options to this in the future to make the icon colors attribute options when I add some of the other variants for this

https://www.figma.com/design/zmKcJ9Xin7ChzyGy6RCFLe/UDS-Runtime-(UI%2FCLI)?node-id=3497-16206&node-type=instance&m=dev

https://www.figma.com/design/zmKcJ9Xin7ChzyGy6RCFLe/UDS-Runtime-(UI%2FCLI)?node-id=2869-4943&node-type=canvas&t=Wk3rHrdAvVP0I36T-0

**Comment by UncleGedd** [comment]
The Pods and Nodes cards on the cluster overview page are looking a bit off. Also the CPU Usage and Memory Usage look like they could use some spacing above and below the text
Screenshot 2024-09-25 at 8 41 45 AM


^ noting this is about ~1280 width on my monitor

File: ui/src/lib/components/StatWidget/ProgressBarWidget.svelte [hunk 1]
```diff
@@ -0,0 +1,25 @@
+<script lang="ts">
```
Comment (UncleGedd) [inline ui/src/lib/components/StatWidget/ProgressBarWidget.svelte:1]: Is this actually a "widget"?

File: ui/src/lib/components/StatWidget/ProgressBarWidget.svelte [hunk 2]
```diff
@@ -0,0 +1,25 @@
...
+    <div class="w-full">
+      <dt class="text-sm font-medium text-gray-500 dark:text-gray-400 truncate">{title}</dt>
+      <dd class="mt-1 text-3xl font-semibold text-gray-900 dark:text-white">
+        {value.toString()}%
+      </dd>
+    </div>
+
+    <ProgressBar size={barSize} {progress} {capacity} {unit} />
+  </div>
+</Card>
```
Comment (UncleGedd) [inline ui/src/lib/components/StatWidget/ProgressBarWidget.svelte:25]: Why does the `ProgressBarWidget` not live in the `ProgressBar` folder?

File: ui/src/lib/components/StatWidget/component.svelte [hunk 3]
```diff
@@ -0,0 +1,14 @@
+<script lang="ts">
+  import IconWidget from './IconWidget.svelte'
+  import ProgressBarWidget from './ProgressBarWidget.svelte'
+  import type { StatType } from './types'
+
+  export let variant: 'with_right_icon' | 'progress_bar'
+  export let props: StatType
+</script>
+
+{#if variant === 'progress_bar' && 'capacity' in props}
```
Comment (UncleGedd) [inline ui/src/lib/components/StatWidget/component.svelte:10]: Is this extra abstraction layer necessary? Does it make sense for the cluster-overview page to just render the `ProgressBarWidget` and the `IconWidget` directly?

**Comment by UncleGedd** [comment]
I think in this case, the widgets are actually cards, thoughts?

**Comment by TristanHoladay** [comment]
posting as well just so you can see what it's like on my screen:

Screenshot from 2024-09-25 07-53-02

File: ui/src/lib/components/StatWidget/ProgressBarWidget.svelte (hunk 1 above)
Comment (BillyFigueroa) [reply ui/src/lib/components/StatWidget/ProgressBarWidget.svelte:1]: Widget is the terminology being used by Flowbite for the "stat" component which includes a bunch more components we have not yet created that will be created in the future

Screenshot 2024-09-25 at 10 48 21 AM

Screenshot 2024-09-25 at 10 49 15 AM

File: ui/src/lib/components/StatWidget/ProgressBarWidget.svelte (hunk 2 above)
Comment (BillyFigueroa) [reply ui/src/lib/components/StatWidget/ProgressBarWidget.svelte:25]: ProgressBar can be used anywhere else. It's not tied to this widget. We can later have another component that uses the ProgressBar that doesn't even include the name "ProgressBar" in it. I could have named this widget something else. Figma toolkit for Flowbite calls this ProgressBar but calls the actual Progress Bar Progress Bars

A widget usually includes multiple components so this is using multiple UI pieces to make a larger component which is basically what a widget is

Screenshot 2024-09-25 at 10 50 48 AM

**Comment by BillyFigueroa** [comment]
> I think in this case, the widgets are actually cards, thoughts?

Figma Flowbite designs calls this a StatWidget. There is a Card component being used inside of the StatWidget which has nothing to do with widgets and is just the structural rounded "card" look to build Cards

**Comment by BillyFigueroa** [comment]
Working on updating the multiple view points to be responsive, then I will do a refactor to name the attributes more like what the Flowbite Figma is using for attributes...ie.

Screenshot 2024-09-25 at 11 08 16 AM

**Comment by BillyFigueroa** [comment]
> The Pods and Nodes cards on the cluster overview page are looking a bit off. Also the CPU Usage and Memory Usage look like they could use some spacing above and below the text Screenshot 2024-09-25 at 8 41 45 AM
> 
> ^ noting this is about ~1280 width on my monitor

Hey @UncleGedd can you pull down latest change and tell me how it looks now. I am curious why the UI never looks that "narrow" on my monitor

**Comment by BillyFigueroa** [comment]
> posting as well just so you can see what it's like on my screen:
> 
> Screenshot from 2024-09-25 07-53-02

@TristanHoladay do you have a few mins to show me this on your own display? My display never shows the cards being that thin

This is full width on a 5k monitor

Screenshot 2024-09-25 at 11 29 07 AM

**Comment by BillyFigueroa** [comment]
> posting as well just so you can see what it's like on my screen:
> 
> Screenshot from 2024-09-25 07-53-02

@TristanHoladay this is a known issue

https://stackoverflow.com/questions/71818458/why-wont-tailwind-find-my-dynamic-class

I will have to add the classes being used to the safeList for dynamically generated classes.

**Review by UncleGedd — APPROVED** [review]
Nice work! Names are bonkers but it's what the Flowbite team is using. @BillyFigueroa is writing up an ADR on naming to ensure both devs and designers are speaking the same language

**Comment by BillyFigueroa** [comment]
> Nice work! Names are bonkers but it's what the Flowbite team is using. @BillyFigueroa is writing up an ADR on naming to ensure both devs and designers are speaking the same language

Adding ADR Link here

https://github.com/defenseunicorns/uds-runtime/pull/401

**Pull request merged**